harness/determined/exec/launch_autohorovod.py

```diff
@@ -0,0 +1,156 @@
+    certs.cli_cert = certs.default_load(
+        master_url=f"http{'s' if use_tls else ''}://{master_addr}:{master_port}"
+    )
+
+    with open(os.environ["DET_LATEST_CHECKPOINT"], "r") as f:
+        latest_checkpoint = json.load(f)
+
+    last_batch = int(os.environ["DET_LATEST_BATCH"])
+
+    use_gpu = distutils.util.strtobool(os.environ.get("DET_USE_GPU", "false"))
+    slot_ids = json.loads(os.environ["DET_SLOT_IDS"])
+    det_rendezvous_port = os.environ["DET_RENDEZVOUS_PORT"]
+    det_trial_unique_port_offset = int(os.environ["DET_TRIAL_UNIQUE_PORT_OFFSET"])
+    det_trial_runner_network_interface = os.environ["DET_TRIAL_RUNNER_NETWORK_INTERFACE"]
+    det_trial_id = os.environ["DET_TRIAL_ID"]
+    det_experiment_id = os.environ["DET_EXPERIMENT_ID"]
+    det_agent_id = os.environ["DET_AGENT_ID"]
+    det_cluster_id = os.environ["DET_CLUSTER_ID"]
+    det_task_token = os.environ["DET_TASK_TOKEN"]
+    trial_seed = int(os.environ["DET_TRIAL_SEED"])
+    trial_run_id = int(os.environ["DET_TRIAL_RUN_ID"])
+    task_id = os.environ["DET_TASK_ID"]
+
+    container_gpus = gpu.get_gpu_uuids_and_validate(use_gpu, slot_ids)
+
+    env = det.EnvContext(
+        master_addr=master_addr,
+        master_port=master_port,
+        use_tls=use_tls,
+        master_cert_file=master_cert_file,
+        master_cert_name=master_cert_name,
+        container_id=container_id,
+        experiment_config=experiment_config,
+        hparams=hparams,
+        latest_checkpoint=latest_checkpoint,
+        last_batch=last_batch,
+        use_gpu=use_gpu,
+        container_gpus=container_gpus,
+        slot_ids=slot_ids,
+        debug=debug,
+        det_rendezvous_port=det_rendezvous_port,
+        det_trial_unique_port_offset=det_trial_unique_port_offset,
+        det_trial_runner_network_interface=det_trial_runner_network_interface,
+        det_trial_id=det_trial_id,
+        det_experiment_id=det_experiment_id,
+        det_agent_id=det_agent_id,
+        det_cluster_id=det_cluster_id,
+        det_task_token=det_task_token,
+        trial_seed=trial_seed,
+        trial_run_id=trial_run_id,
+        task_id=task_id,
+        managed_training=True,
+        test_mode=False,
+        on_cluster=True,
+    )
+
+    logging.info(
+        f"New trial runner in (container {container_id}) on agent {agent_id}: {env.__dict__}."
+    )
+
+    try:
+        storage.validate_config(
+            env.experiment_config["checkpoint_storage"],
+            container_path=constants.SHARED_FS_CONTAINER_PATH,
+        )
+    except Exception as e:
+        logging.error("Checkpoint storage validation failed: {}".format(e))
+        return 1
+
+    jri = json.loads(os.environ["DET_RENDEZVOUS_INFO"])
+    rendezvous_info = det.RendezvousInfo(
+        addrs=jri["addresses"],
+        rank=jri["rank"],
+    )
+
+    hvd_config = horovod.HorovodContext.from_configs(
+        env.experiment_config, rendezvous_info, env.hparams
+    )
+
+    if hvd_config.use:
```

Comment:
I mean, we _probably_ could.  On the other hand, is it possible for us to either totally delete the `EnvContext` object, or keep it as a purely internal object that we create as part of the Generic API?  I'd much prefer not exposing that to users if I could help it.

Like if I look at the current environment variables we read, I see a bunch of junk that seems completely unrelated to the launch layer:

```    "DET_MASTER_ADDR",
    "DET_MASTER_PORT",
    "DET_AGENT_ID",
    "DET_EXPERIMENT_ID",
    "DET_TRIAL_ID",
    "DET_TRIAL_SEED",
    "DET_EXPERIMENT_CONFIG",
    "DET_HPARAMS",
    "DET_LATEST_CHECKPOINT",
    "DET_LATEST_BATCH",
    "DET_TASK_TOKEN",
    "DET_TRIAL_RUN_ID",
    "DET_TASK_ID",
```

And a much smaller list of things that seem definitely necessary for the launch layer:
```
    "DET_TRIAL_RUNNER_NETWORK_INTERFACE",
    "DET_RENDEZVOUS_PORT",
    "DET_RENDEZVOUS_INFO",
```

And maybe a short list of things that seem like they could be useful to both launch layer and the training loop:
```
    "DET_SLOT_IDS",
    "DET_CONTAINER_ID",
    "DET_USE_GPU",
```

I think what I have currently in the launch layer is more out of convenience rather than a correctly drawn line of abstraction :disappointed: